### ISSUE TITLE
[opd, trainer] Supervised distillation loss normalization depends on micro-batch partitioning

### ISSUE THREAD

**Issue by yph22** [verl-project/verl#7200]
### System Info

verl:0.9.0.dev

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The supervised/direct on-policy distillation loss is normalized independently
inside every micro-batch because `config.global_batch_info` is empty when
`distillation_loss()` is called.

As a result, the accumulated loss and gradients depend on how the same mini-batch
is partitioned into micro-batches. With static micro-batch size 1, the effective
objective becomes a sum of per-sequence token means instead of the configured
global token mean.

This is reproducible with the following target configuration:

```yaml
distillation:
  enabled: true
  distillation_loss:
    loss_mode: forward_kl_topk
    use_task_rewards: false
    use_policy_gradient: false

actor_rollout_ref:
  actor:
    loss_agg_mode: token-mean
    use_dynamic_bsz: false
    ppo_micro_batch_size_per_gpu: 1
```

### Expected behavior

For `loss_agg_mode=token-mean`, let:

- \(D\) be the data-parallel size;
- \(N\) be the number of valid response tokens in the global mini-batch;
- \(T_{r,m}\) be the tokens on DP rank \(r\), micro-batch \(m\);
- \(\ell_t\) be the per-token distillation loss.

The intended global objective is:

\[
L =
\frac{1}{N}
\sum_{r=1}^{D}
\sum_m
\sum_{t\in T_{r,m}}\ell_t.
\]

Because DDP averages gradients across ranks, each rank/micro-batch should
backpropagate:

\[
L_{r,m} =
\frac{D}{N}
\sum_{t\in T_{r,m}}\ell_t.
\]

Therefore, splitting one mini-batch into one or multiple micro-batches must not
change either the accumulated scalar loss or the parameter gradients.

`agg_loss()` already implements this behavior when it receives
`dp_size=D` and `batch_num_tokens=N`:

```python
loss = verl_F.masked_sum(loss_mat, loss_mask) / batch_num_tokens * dp_size
```

## Actual behavior

The direct distillation branch calls:

```python
distillation_loss = agg_loss(
    loss_mat=distillation_losses,
    loss_mask=response_mask,
    loss_agg_mode=loss_agg_mode,
    **config.global_batch_info,
)
```

However, `ActorConfig.global_batch_info` is initialized as an empty dictionary.
For supervised top-k OPD, `ppo_loss()` is skipped, so nothing populates that
dictionary before `agg_loss()` is called.

Consequently, `agg_loss()` receives its defaults:

```python
dp_size = 1
batch_num_tokens = None
```

and falls back to:

```python
batch_num_tokens = loss_mask.sum()
loss = masked_sum(loss_mat, loss_mask) / batch_num_tokens
```

This is a local micro-batch mean. The FSDP engine then performs backward once per
micro-batch and accumulates these independently normalized gradients:

```python
for micro_batch in micro_batches:
    loss, meta_info = self.forward_step(...)
    loss.backward()
```

The effective objective becomes:

\[
\widetilde L =
\frac{1}{D}
\sum_{r=1}^{D}
\sum_m
\left(
\frac{1}{N_{r,m}}
\sum_{t\in T_{r,m}}\ell_t
\right),
\]

which depends on the micro-batch partition.

**Comment by wuxibin89** [comment]
Yeah, this is a bug but not very severe, accumulates gradient on local micro-batch mean usually works as well. To fix, we can add global batch info for distillation_loss:
https://github.com/verl-project/verl/blob/main/verl/workers/utils/losses.py#L64-L68

**Comment by zyang6** [comment]
> Yeah, this is a bug but not very severe, accumulates gradient on local micro-batch mean usually works as well. To fix, we can add global batch info for distillation_loss: https://github.com/verl-project/verl/blob/main/verl/workers/utils/losses.py#L64-L68

Thanks for the fix in #7225!

We @alwaysyiyu  recently hit the same bug. While looking into it, we also had a question about the `batch_num_tokens is None` fallback in `agg_loss` — opened a short discussion here: #7321  Would love to discuss this with you!
